### PR TITLE
Monomorphization with Call Graph

### PR DESCRIPTION
**Pull request by ScottCarda-MS**
Uses the concrete call graph to do monomorphization. The rewrite step is now only responsible for generating the callables based on the nodes in the graph.

### CODE REVIEW

**Comment by ScottCarda-MS** [comment]
Will be adding unit tests for the concrete call graph in this PR.

**Comment by ScottCarda-MS** [comment]
Unit tests for the concrete call graph were moved to #597 and are now in the feature branch.

File: src/QsCompiler/Compiler/CompilationLoader.cs [hunk 1]
```diff
@@ -568,6 +562,12 @@ public CompilationLoader(SourceLoader loadSources, ReferenceLoader loadReference
                 steps.Add((rewriteStep.Priority, () => this.ExecuteAsAtomicTransformation(rewriteStep, ref this.compilationStatus.PreEvaluation)));
             }
 
+            if (this.config.IsExecutable && !this.config.SkipMonomorphization)
```
Comment (cesarzc) [inline src/QsCompiler/Compiler/CompilationLoader.cs:565]: Was this moved after `GenerateFunctorSupport` for some reason?

Comment (ScottCarda-MS) [reply src/QsCompiler/Compiler/CompilationLoader.cs:565]: Yes Monomorphization is going to be the last rewrite step processed before external rewrite steps now. I guess the order in this file is not super important, as the priority number is actually the thing that determines the order, but having the order here match that I think makes things more consistent.

Comment (ScottCarda-MS) [reply src/QsCompiler/Compiler/CompilationLoader.cs:565]: The reason Monomorphization has to be processed after functor support is that we don't have a way to deal with auto-generated functor specializations that have not been generated yet when building the concrete call graph, so the solution is to run functor generation first to make sure all of the functor specializations are generated and present.

Comment (cesarzc) [reply src/QsCompiler/Compiler/CompilationLoader.cs:565]: Thanks for the explanation Scott! Yes, I like the consistency.

**Review by cesarzc — APPROVED** [review]
Looks good to me! The call graph really makes this simpler

**Pull request merged**